File: Samples/Model-Router/TypeScript/model-router-multiturn.ts
import ModelClient from "@azure-rest/ai-inference";
import { AzureKeyCredential } from "@azure/core-auth";
import * as readline from 'readline';
import "dotenv/config";

type ChatMessage = {
    role: "system" | "user" | "assistant";
    content: string;
};

const { AZURE_INFERENCE_SDK_ENDPOINT: endpoint, AZURE_INFERENCE_SDK_KEY: key } = process.env;

// Trim conversation history to keep system message plus the last N user/assistant turns.
function trimHistory(messages: ChatMessage[], maxHistoryTurns: number): ChatMessage[] {
    if (!messages.length) {
        return messages;
    }

    // Always preserve the first message if it is the system message
    const head = messages.slice(0, 1);
    const body = messages.slice(1);

    // Count pairs in body. We'll keep the last 2*maxHistoryTurns messages
    const keep = maxHistoryTurns * 2;
    return [...head, ...body.slice(-keep)];
}

export async function main(): Promise<void> {
    if (!endpoint || !key) {
        throw new Error("AZURE_INFERENCE_SDK_ENDPOINT and AZURE_INFERENCE_SDK_KEY environment variables must be set");
    }

    const client = ModelClient(endpoint, new AzureKeyCredential(key));
    
    const messages: ChatMessage[] = [
        { role: "system", content: "You are a helpful assistant." }
    ];

    const maxHistoryTurns = 12;

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    console.log("Interactive chat started. Type 'exit' or 'quit' to stop.\n");

    const askQuestion = (question: string): Promise<string> => {
        return new Promise((resolve) => {
            rl.question(question, (answer) => resolve(answer));
        });
    };

    try {
        while (true) {
            const userInput = (await askQuestion("You: ")).trim();
            const isExitCommand = ["exit", "quit"].includes(userInput.toLowerCase());

            if (!userInput || isExitCommand) {
                if (isExitCommand) console.log("Goodbye.");
                break;
            }

            messages.push({ role: "user", content: userInput });

            const trimmedMessages = trimHistory(messages, maxHistoryTurns);

            const requestBody = {
                messages: trimmedMessages,
                max_tokens: 8192,
                temperature: 0.7,
                top_p: 0.95,
                frequency_penalty: 0,
                presence_penalty: 0,
                model: "model-router",
            };

            const response = await client.path("/chat/completions").post({
                body: requestBody,
            });

            const { body } = response as any;
            const assistantContent = body?.choices?.[0]?.message?.content;

            console.log(`Model chosen by the router: ${body?.model ?? "(unknown)"}`);
            console.log(`Assistant: ${assistantContent}`);
            console.log(`Usage:`, body?.usage);
            console.log();

            messages.push({ role: "assistant", content: assistantContent });
        }
    } finally {
        rl.close();
    }
}

main().catch((err) => {
    console.error("The sample encountered an error:", err);
    process.exit(1);
});

File: Samples/Model-Router/Python/model-router-multiturn.py
import os
import sys
from typing import List, Dict

from openai import AzureOpenAI
from dotenv import load_dotenv


def main():
    """Run an interactive, multi-turn console chat using Azure OpenAI model router.

    - Maintains conversation history (system + rolling window of recent turns)
    - Sends full history each turn for context
    - Type 'exit' or 'quit' to end the session
    """

    load_dotenv()

    subscription_key = os.getenv("subscription_key")
    endpoint = os.getenv("endpoint")
    deployment = os.getenv("deployment")
    api_version = os.getenv("api_version")

    # System prompt can be customized via env
    system_prompt = os.getenv("SYSTEM_PROMPT", "You are a helpful assistant.")

    client = AzureOpenAI(
        api_version=api_version,
        azure_endpoint=endpoint,
        api_key=subscription_key,
    )

    # Conversation state: always begin with system message
    messages: List[Dict[str, str]] = [
        {"role": "system", "content": system_prompt},
    ]

    # Controls how many turns (user+assistant pairs) to keep in context
    max_history_turns = int(os.getenv("MAX_HISTORY_TURNS", "12"))

    print("Interactive chat started. Type 'exit' or 'quit' to stop.\n")

    while True:
        try:
            user_input = input("You: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nExiting.")
            break

        if not user_input:
            continue
        if user_input.lower() in {"exit", "quit"}:
            print("Goodbye.")
            break

        # Append user turn
        messages.append({"role": "user", "content": user_input})

        # Trim history to keep context small while preserving the system message
        trimmed = trim_history(messages, max_history_turns=max_history_turns)

        # Call the router deployment
        response = client.chat.completions.create(
            messages=trimmed,
            max_tokens=int(os.getenv("MAX_OUTPUT_TOKENS", "10000")),
            temperature=float(os.getenv("TEMPERATURE", "0.7")),
            top_p=float(os.getenv("TOP_P", "0.95")),
            frequency_penalty=float(os.getenv("FREQ_PENALTY", "0.0")),
            presence_penalty=float(os.getenv("PRES_PENALTY", "0.0")),
            model=deployment,
        )

        assistant_message = response.choices[0].message.content or ""
        routed_model = getattr(response, "model", "<unknown>")

        print(f"\n[model: {routed_model}]\nAssistant: {assistant_message}\n")

        # Append assistant turn to the full history
        messages.append({"role": "assistant", "content": assistant_message})


def trim_history(messages: List[Dict[str, str]], max_history_turns: int) -> List[Dict[str, str]]:
    """Keep system message plus the last N user/assistant turns.

    messages: full history including system at index 0
    max_history_turns: number of recent user+assistant pairs to retain
    """
    if not messages:
        return messages

    # Always preserve the first message if it is the system message
    head = messages[0:1]
    body = messages[1:]

    # Count pairs in body. We'll keep the last 2*max_history_turns messages
    keep = max_history_turns * 2
    if len(body) <= keep:
        return head + body
    else:
        return head + body[-keep:]


if __name__ == "__main__":
    main()